### ISSUE TITLE
How to run NN and model on novel input data?

### ISSUE THREAD

**Issue by aindilis** [declare-lab/dialogue-understanding#8]
Hi,

This looks to be a great system!

First, I am relatively new to NNs, and I've noticed that often times models are not distributed with code, nor is there an easy entry point for using the model on novel data. I am assuming I am missing something involved in the process, but I don't know what that would be.  Could you please explain to me why oftentimes ML codebases don't come with pre-trained models or evaluation capabilities?

Second, I have used the following command to begin training a model:

```cd roberta-end-to-end && python train.py --dataset dailydialog --classify act --cls-model lstm --residual --batch-size 2```

I don't know Python very well. I am wondering is there an easy way to use this trained model, when completed, in order to process texts? I can definitely preprocess the text into any format using Perl. I just don't know how to run/evaluate the NN on specific input data for use in classifying texts (in my case I'm trying to extract commissives from multi-party dialogues (however it looked like these dialogues were two-person dialogues, so perhaps I should have trained the utterance model)).

Thank you for making this software available, it may be a great help to us,

Andrew

**Comment by aindilis** [comment]
Another thing to note is I trained the model and it didn't appear to save it anywhere.  Is there a document which talks about defaults and expectations when it comes to work in ML/NN/DL where one might learn such?  Also, I would like to know how to properly save it, is it just a matter of passing the output_dir argument?

**Comment by deepanwayx** [comment]
I will divide my answer into two comments. This one is about saving the trained models and how to use them later.

Recent transformer models are often quite large in size, so we couldn't upload the trained models to GitHub. We also experimented with a lot of configurations in this project, so it was impractical to upload all the models. 

However, saving a model is quite easy and can be done with the following change:

```
torch.save(model.state_dict(), "saved/dialog_bert_transformer_best_val.pt")
```
after line 263 in `roberta-end-to-end/train.py` inside the if condition

Then if you run the training script, 
`cd roberta-end-to-end && python train.py --dataset dailydialog --classify act --cls-model lstm --residual --batch-size 2`

the best model weights according to the validation loss would be saved in the `saved` directory.

Once the model training is completed, you can load the trained model (perhaps in your inference script) and prepare the model for inference as follows:

```
model = DialogBertTransformer(D_h, classification_model, transformer_model, transformer_mode, n_classes, context_attention, attention, residual)
model.load_state_dict(torch.load("saved/dialog_bert_transformer_best_val.pt"))
model.eval()
``` 

Remember to use the same values of `D_h, classification_model` and other arguments as you did during training. 

You can use the same `train_or_eval_model` function in the [training script](https://github.com/declare-lab/dialogue-understanding/blob/master/roberta-end-to-end/train.py) to obtain the predictions for your dataset in the `preds` variable. The inner elements in `preds` for which the inner elements in `mask` are non-zero are the ones you care about, as the others were [zero-padded elements](https://github.com/declare-lab/dialogue-understanding/blob/master/roberta-end-to-end/train.py#L98). 

You can read more about saving and loading models [here](https://pytorch.org/tutorials/beginner/saving_loading_models.html).

**Comment by deepanwayx** [comment]
To use your own data for inference, prepare your data in the tab-separated format as we have written in the Data Format section of [ReadMe](https://github.com/declare-lab/dialogue-understanding). The files all follow one dialogue in each line format. The loss_masks file should be all 1. You can create a dummy labels file if you do not have labels for your dataset.

Then you can create the `test_loader` with your own files in the [script](https://github.com/declare-lab/dialogue-understanding/blob/master/roberta-end-to-end/train.py#L58) and use them for prediction.

You also mentioned that your dataset has multi-party dialogues, but the DailyDialog dataset (which has the commissive act label) is only two-party. So you have to exclusively use `--cls-model lstm` for training. The other dialogrnn classification model uses speaker information and would throw an error if you pass speaker ids during inference which were not used in training. 

The `--cls-model lstm` uses only the content of the utterance for prediction. It does not use speaker information, so you can use the DailyDialog trained model for your dataset.
